esp-config/src/generate.rs

```diff
@@ -0,0 +1,138 @@
+/// the config has been parsed, this function will emit snake case cfg's
+/// _without_ the prefix which can be used in the dependant crate. After that,
+/// it will create a markdown table in the `OUT_DIR` under the name
+/// `{prefix}_config_table.md` where prefix has also been converted
+/// to snake case. This can be included in crate documentation to outline the
+/// available configuration options for the crate.
+///
+/// The tuple ordering for the`config` array is key, default, description.
+///
+/// Unknown keys with the supplied prefix will cause this function to panic.
+pub fn generate_config(prefix: &str, config: &[(&str, Value, &str)]) {
+    let mut prefix = screaming_snake_case(prefix);
+
+    let mut doc_table = String::from(TABLE_HEADER);
+    for (key, default, desc) in config {
+        let key = screaming_snake_case(key);
+        let default = default.as_string();
+        writeln!(doc_table, "|**{prefix}_{key}**|{desc}|{default}|").unwrap();
+    }
+
+    // only rebuild if build.rs changed. Otherwise Cargo will rebuild if any
+    // other file changed.
+    println!("cargo:rerun-if-changed=build.rs");
+
+    // Generate the template for the config
+    let mut configs = HashMap::new();
+    for (name, default, _) in config {
+        configs.insert(snake_case(name), default.clone());
+
+        // Rebuild if config envvar changed.
+        // Note this is currently buggy and requires a clean build - PR is pending https://github.com/rust-lang/cargo/pull/14058
+        println!("cargo:rerun-if-env-changed={name}");
+    }
+
+    // Try and capture input from the environment
+    for (var, value) in env::vars() {
+        if let Some(name) = var.strip_prefix(&format!("{prefix}_")) {
+            let name = snake_case(name);
+            // TODO should this be lowered to a warning for unknown configuration options?
+            // we could instead mark this as unknown and _not_ emit
+            // println!("cargo:rustc-check-cfg=cfg({})", name);
+            let Some(cfg) = configs.get_mut(&name) else {
+                panic!("Unknown env var {name}")
+            };
+
+            cfg.parse(&value)
+                .expect(&format!("Invalid value for env var {name}: {value}"));
+        }
+    }
+
+    // emit cfgs and set envs
+    for (name, value) in configs.into_iter() {
+        println!("cargo:rustc-check-cfg=cfg({})", name);
+        match value {
+            Value::Bool(val) if val == true => println!("cargo:rustc-cfg={name}"),
+            _ => {}
+        }
+
+        // values that haven't been seen will be output here with the default value
+        println!(
+            "cargo:rustc-env={}={}",
+            format!("{prefix}_{}", screaming_snake_case(&name)),
+            value.as_string()
+        );
+    }
+
+    // convert to snake case
+    prefix.make_ascii_lowercase();
+
+    let out_dir = PathBuf::from(env::var_os("OUT_DIR").unwrap());
+    let out_file = out_dir
```

Comment:
It would be nice to dump the active configuration to `./target/esp-{crate-name}.md`, too. I know the cargo output can be found somewhere deep but an rtic-style "here's an easy to find set of files for you" could be useful IMO.